core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -682,266 +706,166 @@ private void syncRemoteToLocal() throws InterruptedException {
-        nsConfig.getNamespace(),
-        docConfig.getDocumentId(),
-        remoteChangeEvent.getOperationType().toString()));
+    SyncAction action = null;
+    SyncMessage message = null;
 
-    final DocumentVersionInfo currentRemoteVersionInfo;
+    docConfig.getLock().readLock().lock();
     try {
-      currentRemoteVersionInfo = DocumentVersionInfo
-          .getRemoteVersionInfo(remoteChangeEvent.getFullDocument());
-    } catch (final Exception e) {
-      emitError(docConfig,
-          String.format(
-              Locale.US,
-              "t='%d': syncRemoteChangeEventToLocal ns=%s documentId=%s got a remote "
-                  + "document that could not have its version info parsed "
-                  + "; dropping the event, and desyncing the document",
-              logicalT,
-              nsConfig.getNamespace(),
-              docConfig.getDocumentId()));
-      return desyncDocumentsFromRemote(nsConfig.getNamespace(), docConfig.getDocumentId());
-    }
-
-
-    if (currentRemoteVersionInfo.hasVersion()
-        && currentRemoteVersionInfo.getVersion().getSyncProtocolVersion() != 1) {
-      emitError(docConfig,
-              String.format(
-                      Locale.US,
-                      "t='%d': syncRemoteChangeEventToLocal ns=%s documentId=%s got a remote "
-                              + "document with an unsupported synchronization protocol version "
-                              + "%d; dropping the event, and desyncing the document",
-                      logicalT,
-                      nsConfig.getNamespace(),
-                      docConfig.getDocumentId(),
-                      currentRemoteVersionInfo.getVersion().getSyncProtocolVersion()));
-      return desyncDocumentsFromRemote(nsConfig.getNamespace(), docConfig.getDocumentId());
-    }
-
-    // ii. If the version info for the unprocessed change event has the same GUID as the local
-    //     document version GUID, and has a version counter less than or equal to the local
-    //     document version version counter, drop the event, as it implies the event has already
-    //     been applied to the local collection.
-    if (docConfig.hasCommittedVersion(currentRemoteVersionInfo)) {
-      // Skip this event since we generated it.
-      logger.debug(String.format(
-          Locale.US,
-          "t='%d': syncRemoteChangeEventToLocal ns=%s documentId=%s remote change event was "
-              + "generated by us; dropping the event",
-          logicalT,
-          nsConfig.getNamespace(),
-          docConfig.getDocumentId()));
-      return null;
-    }
+      if (docConfig.hasUncommittedWrites() && docConfig.getLastResolution() == logicalT) {
+        action = SyncAction.WAIT;
+        message = SyncMessage.SIMULTANEOUS_WRITES_MESSAGE;
+      }
 
+      if (logger.isDebugEnabled()) {
+        logger.debug(String.format(
+            Locale.US,
+            SyncMessage.BASE_MESSAGE.concat(" processing remote operation='%s'"),
```

Comment:
I still think it's a bit hard to read, because it's unclear what the other arguments in `String.format` are referring to unless you actually look at the `BASE_MESSAGE` string. In any case, this is pretty nitty and not that big of a deal